File: frontend/Concord/src/api/messagesService.js
import { request } from './httpClient';
import { buildPagingQuery } from './pagination';

export async function getMessages(serverId, channelId, { page, pageSize } = {}) {
  return request(`Servers/${serverId}/Channels/${channelId}/Messages`, {
    query: buildPagingQuery(page, pageSize),
  });
}

export async function sendMessage(serverId, channelId, { Content, AttachmentUrl = null, ReplyToMessageId = null }) {
  return request(`Servers/${serverId}/Channels/${channelId}/Messages`, {
    method: 'POST',
    body: { Content, AttachmentUrl, ReplyToMessageId },
  });
}

export async function editMessage(serverId, channelId, messageId, content) {
  return request(`Servers/${serverId}/Channels/${channelId}/Messages/${messageId}`, {
    method: 'PUT',
    body: { Content: content },
  });
}

export async function deleteMessage(serverId, channelId, messageId) {
  await request(`Servers/${serverId}/Channels/${channelId}/Messages/${messageId}`, { method: 'DELETE' });
}

export async function toggleReaction(serverId, channelId, messageId, emoji) {
  return request(`Servers/${serverId}/Channels/${channelId}/Messages/${messageId}/Reactions`, {
    method: 'POST',
    body: { Emoji: emoji },
  });
}

export async function pinMessage(serverId, channelId, messageId) {
  return request(`Servers/${serverId}/Channels/${channelId}/Messages/${messageId}:Pin`, { method: 'POST' });
}

export async function unpinMessage(serverId, channelId, messageId) {
  return request(`Servers/${serverId}/Channels/${channelId}/Messages/${messageId}:Unpin`, { method: 'POST' });
}

export async function getPinnedMessages(serverId, channelId) {
  return request(`Servers/${serverId}/Channels/${channelId}/Messages/Pinned`);
}

export async function forwardMessage(serverId, channelId, messageId, { targetChannelId = null, targetConversationId = null }) {
  return request(`Servers/${serverId}/Channels/${channelId}/Messages/${messageId}/Forward`, {
    method: 'POST',
    body: { TargetChannelId: targetChannelId, TargetConversationId: targetConversationId },
  });
}

File: mobile/concord/lib/api/messages_service.dart
import 'api_client.dart';
import 'models/message_response.dart';
import 'models/paged_result.dart';

class MessagesService {
  MessagesService(this._client);

  final ApiClient _client;

  Future<PagedResult<MessageResponse>> listMessages(
    String serverId,
    String channelId, {
    int page = 1,
    int pageSize = 30,
  }) async {
    final data = await _client.get(
      '/Servers/$serverId/Channels/$channelId/Messages',
      query: {'page': page, 'pageSize': pageSize},
    );
    return PagedResult.fromJson(data as Map<String, dynamic>, MessageResponse.fromJson);
  }

  Future<MessageResponse> sendMessage(
    String serverId,
    String channelId, {
    required String content,
    String? attachmentUrl,
    String? replyToMessageId,
  }) async {
    final data = await _client.post(
      '/Servers/$serverId/Channels/$channelId/Messages',
      body: {
        'Content': content,
        'AttachmentUrl': attachmentUrl,
        'ReplyToMessageId': replyToMessageId,
      },
    );
    return MessageResponse.fromJson(data as Map<String, dynamic>);
  }

  Future<MessageResponse> editMessage(
    String serverId,
    String channelId,
    String messageId, {
    required String content,
  }) async {
    final data = await _client.put(
      '/Servers/$serverId/Channels/$channelId/Messages/$messageId',
      body: {'Content': content},
    );
    return MessageResponse.fromJson(data as Map<String, dynamic>);
  }

  Future<bool> deleteMessage(String serverId, String channelId, String messageId) async {
    final data = await _client.delete('/Servers/$serverId/Channels/$channelId/Messages/$messageId');
    return data as bool;
  }

  Future<MessageResponse> toggleReaction(
    String serverId,
    String channelId,
    String messageId,
    String emoji,
  ) async {
    final data = await _client.post(
      '/Servers/$serverId/Channels/$channelId/Messages/$messageId/Reactions',
      body: {'Emoji': emoji},
    );
    return MessageResponse.fromJson(data as Map<String, dynamic>);
  }

  Future<MessageResponse> pinMessage(String serverId, String channelId, String messageId) async {
    final data = await _client.post('/Servers/$serverId/Channels/$channelId/Messages/$messageId:Pin');
    return MessageResponse.fromJson(data as Map<String, dynamic>);
  }

  Future<MessageResponse> unpinMessage(String serverId, String channelId, String messageId) async {
    final data = await _client.post('/Servers/$serverId/Channels/$channelId/Messages/$messageId:Unpin');
    return MessageResponse.fromJson(data as Map<String, dynamic>);
  }

  Future<List<MessageResponse>> getPinnedMessages(String serverId, String channelId) async {
    final data = await _client.get('/Servers/$serverId/Channels/$channelId/Messages/Pinned');
    return (data as List<dynamic>)
        .map((e) => MessageResponse.fromJson(e as Map<String, dynamic>))
        .toList();
  }

  /// Forwards a message sourced from this channel into another channel or DM conversation.
  /// Exactly one of [targetChannelId]/[targetConversationId] must be set — mirrors the backend's
  /// `ForwardMessageRequest`. The response shape depends on which target was chosen (a
  /// `MessageResponse` or a `DirectMessageResponse`), which callers don't need: the newly created
  /// message reaches its destination thread over that thread's own SignalR hub, same as any other
  /// send, so this just needs to succeed or throw.
  Future<void> forwardMessage(
    String serverId,
    String channelId,
    String messageId, {
    String? targetChannelId,
    String? targetConversationId,
  }) async {
    await _client.post(
      '/Servers/$serverId/Channels/$channelId/Messages/$messageId/Forward',
      body: {
        'TargetChannelId': targetChannelId,
        'TargetConversationId': targetConversationId,
      },
    );
  }
}
